File: mailbox-android/src/main/java/org/briarproject/mailbox/core/system/AndroidSystemModule.java
/*
 *     Briar Mailbox
 *     Copyright (C) 2021-2022  The Briar Project
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package org.briarproject.mailbox.core.system;

import org.briarproject.mailbox.core.event.EventExecutor;
import org.briarproject.mailbox.core.lifecycle.LifecycleManager;

import java.util.concurrent.Executor;
import java.util.concurrent.RejectedExecutionHandler;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledThreadPoolExecutor;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public class AndroidSystemModule {

	private final ScheduledExecutorService scheduledExecutorService;

	public AndroidSystemModule() {
		// Discard tasks that are submitted during shutdown
		RejectedExecutionHandler policy =
				new ScheduledThreadPoolExecutor.DiscardPolicy();
		scheduledExecutorService = new ScheduledThreadPoolExecutor(1, policy);
	}

	@Provides
	@Singleton
	ScheduledExecutorService provideScheduledExecutorService(
			LifecycleManager lifecycleManager) {
		lifecycleManager.registerForShutdown(scheduledExecutorService);
		return scheduledExecutorService;
	}

	@Provides
	@Singleton
	AndroidExecutor provideAndroidExecutor(
			AndroidExecutorImpl androidExecutor) {
		return androidExecutor;
	}

	@Provides
	@Singleton
	@EventExecutor
	Executor provideEventExecutor(AndroidExecutor androidExecutor) {
		return androidExecutor::runOnUiThread;
	}

	@Provides
	@Singleton
	AndroidWakeLockManager provideWakeLockManager(
			AndroidWakeLockManagerImpl wakeLockManager) {
		return wakeLockManager;
	}
}

File: mailbox-android/src/main/java/org/briarproject/mailbox/android/system/AndroidSystemModule.kt
package org.briarproject.mailbox.android.system

import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import org.briarproject.mailbox.android.api.system.AndroidExecutor
import org.briarproject.mailbox.android.api.system.AndroidWakeLockManager
import org.briarproject.mailbox.core.lifecycle.LifecycleManager
import java.util.concurrent.Executor
import java.util.concurrent.RejectedExecutionHandler
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.ThreadPoolExecutor
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
internal class AndroidSystemModule {

    private var scheduledExecutorService: ScheduledExecutorService

    init {
        // Discard tasks that are submitted during shutdown
        val policy: RejectedExecutionHandler = ThreadPoolExecutor.DiscardPolicy()
        scheduledExecutorService = ScheduledThreadPoolExecutor(1, policy)
    }

    @Provides
    @Singleton
    fun provideScheduledExecutorService(
        lifecycleManager: LifecycleManager,
    ): ScheduledExecutorService {
        lifecycleManager.registerForShutdown(scheduledExecutorService)
        return scheduledExecutorService
    }

    @Provides
    @Singleton
    fun provideWakeLockManager(
        wakeLockManager: AndroidWakeLockManagerImpl,
    ): AndroidWakeLockManager {
        return wakeLockManager
    }

    @Provides
    @Singleton
    fun provideAndroidExecutor(androidExecutor: AndroidExecutorImpl): AndroidExecutor {
        return androidExecutor
    }

    @Provides
    @Singleton
    fun provideEventExecutor(androidExecutor: AndroidExecutor): Executor {
        return Executor {
            androidExecutor.runOnUiThread(it)
        }
    }

}
